README.md
## Lil Tycoon!

Github Pages: https://jacoby-y.github.io/lil-tycoon/

changelog.md
## v0.4.3
- Fixed images not loading

## v0.4.2
- Removed debug things

## v0.4.1
- Made game view scale fully

## v0.4
- Minimum viable product

## v0.3
- Reworking entire ECS System

## v0.2
- Reworked code
- Added orb value multipliers
- Extended functionality for entity system

## v0.1
- Initial commit

src/game.js
import { writable, get } from "svelte/store";
import draw from "./utils/draw.js";

/** @type {HTMLCanvasElement}*/
export const canvas = writable(null);
let w = 0, h = 0;
canvas.subscribe((v)=>{
	if (v != null) {
		w = v.width;
		h = v.height;
	}
})

export const timer = writable(0);
setInterval(() => { 
	timer.update((v)=>(v+1)%30); 
	job_m.update();
}, 1000/30);

export const job_m = {
	jobs: [],
	add(func, ticks) {
		this.jobs.push({ func, ticks });
	},
	update() {
		for (let i = this.jobs.length-1; i >= 0; i--) {
			const job = this.jobs[i];
			job.ticks--;
			if (job.ticks <= 0) {
				job.func();
				this.jobs.splice(i, 1);
			}
		}
	}
}

export const mouse = {
	x: -1, y: -1,
	hover: false,
	down: false,
}

export const player = {
	x: 500, y: 300, size: 30,
	update() {
		const [ pw, ph ] = [ this.size, this.size ];
		
		this.x += contr.x*7;
		this.y += contr.y*7;
		if (this.x < 0) this.x = 0;
		if (this.y < 0) this.y = 0;
		if (this.x+pw > w) this.x = w-pw;
		if (this.y+ph > h) this.y = h-ph;
	},
	draw() { draw.rect(this.x, this.y, this.size, this.size, "white"); }
}
export const contr = {
	a: false, d: false,
	w: false, s: false,
	e: false,
	get x() {
		const horz = (this.a ? -1 : 0) + (this.d ? 1 : 0);
		const vert = (this.w ? -1 : 0) + (this.s ? 1 : 0);
		if (horz != 0 && vert != 0)
			return 0.707 * horz;
		return horz;
	},
	get y() {
		const horz = (this.a ? -1 : 0) + (this.d ? 1 : 0);
		const vert = (this.w ? -1 : 0) + (this.s ? 1 : 0);
		if (horz != 0 && vert != 0)
			return 0.707 * vert;
		return vert;
	}
}

export const drop_speeds = {
	drop_group_1: 120,
	drop_group_2: 120,
}

rollup.config.js
import svelte from 'rollup-plugin-svelte';
import commonjs from '@rollup/plugin-commonjs';
import resolve from '@rollup/plugin-node-resolve';
import json from '@rollup/plugin-json'; 
import livereload from 'rollup-plugin-livereload';
import { terser } from 'rollup-plugin-terser';
import css from 'rollup-plugin-css-only';

const production = !process.env.ROLLUP_WATCH;

function serve() {
	let server;

	function toExit() {
		if (server) server.kill(0);
	}

	return {
		writeBundle() {
			if (server) return;
			server = require('child_process').spawn('npm', ['run', 'start', '--', '--dev'], {
				stdio: ['ignore', 'inherit', 'inherit'],
				shell: true
			});

			process.on('SIGTERM', toExit);
			process.on('exit', toExit);
		}
	};
}

export default {
	input: 'src/main.js',
	output: {
		sourcemap: true,
		format: 'iife',
		name: 'app',
		file: 'public/build/bundle.js'
	},
	plugins: [
		svelte({
			compilerOptions: {
				// enable run-time checks when not in production
				dev: !production
			},
		}),
		json({ compact: true }),
		// we'll extract any component CSS out into
		// a separate file - better for performance
		css({ output: 'bundle.css' }),

		// If you have external dependencies installed from
		// npm, you'll most likely need these plugins. In
		// some cases you'll need additional configuration -
		// consult the documentation for details:
		// https://github.com/rollup/plugins/tree/master/packages/commonjs
		resolve({
			browser: true,
			dedupe: ['svelte']
		}),
		commonjs(),

		// In dev mode, call `npm run start` once
		// the bundle has been generated
		!production && serve(),

		// Watch the `public` directory and refresh the
		// browser on changes when not in production
		!production && livereload('public'),

		// If we're building for production (npm run build
		// instead of npm run dev), minify
		production && terser()
	],
	watch: {
		clearScreen: false
	}
};


src/utils/draw.js
/** @type {CanvasRenderingContext2D}*/
let ctx;
const set_ctx = (_ctx)=>{ 
	ctx = _ctx; 
}
const circle = (x,y, r, color, fill=true, l_width=1)=>{
	if (ctx == undefined) return;
	if (fill) ctx.fillStyle = color;
	else { ctx.strokeStyle = color; ctx.lineWidth = l_width; }
	ctx.beginPath();
	ctx.arc(x,y, r, 0, Math.PI*2);
	if (fill) ctx.fill();
	else ctx.stroke();
}
const rect = (x,y, w,h, color, fill=true, l_width=1)=>{
	if (ctx == undefined) return;
	if (fill) ctx.fillStyle = color;
	else { ctx.strokeStyle = color; ctx.lineWidth = l_width; }
	if (fill) ctx.fillRect(x,y, w,h);
	else ctx.strokeRect(x,y, w,h);
}
const image = (img, sx=0, sy=0, sWidth=100, sHeight=100)=>{
	ctx.drawImage(img, sx, sy, sWidth, sHeight);
}

const new_path = (...pts)=>{
	let build = [];
	for (let i = 0; i < pts.length; i += 2) {
		const pt1 = pts[i];
		const pt2 = pts[i+1];
		build.push({
			x: pt1,
			y: pt2,
		});
	}
	return build;
}
const path = (points=[], color="black", fill=true, l_width=1)=>{
	if (fill) ctx.fillStyle = color;
	else { ctx.strokeStyle = color; ctx.lineWidth = l_width; }
	ctx.beginPath();
	ctx.moveTo(points[0].x, points[1].y);
	for (let i = 0; i < points.length; i++) {
		const pt = points[i];
		ctx.lineTo(pt.x, pt.y);
	}
	ctx.closePath();
	if (fill) ctx.fill();
	else ctx.stroke();
}

const transform = (x=0,y=0, ang=0)=>{
	ctx.translate(x,y);
	ctx.rotate(ang);
}
const reset = ()=>{
	ctx.setTransform(1, 0, 0, 1, 0, 0);
}

export default {
	set_ctx,
	circle,
	rect,
	image,
	new_path,
	path,
	transform,
	reset
};

src/utils/utils.js
/** @param {Number} num */
export const exp = (pow=1)=> (10 ** pow)
export const round = (num, pow=1)=> Math.round(num * exp(pow))/exp(pow);
export const floor = (num, pow=1)=> Math.floor(num * exp(pow))/exp(pow);
export const ceil = (num, pow=1)=> Math.ceil(num * exp(pow))/exp(pow);
export const toExp = (num, place)=>{
  let pow = Math.floor(Math.log10(num));
	place = 10**place;
  return (Math.floor(num/(10**pow)*place)/place + `e${pow}`);
}
export const sci = (num, place=1)=> num >= 1000 ? toExp(num, place).replace("+", "") : round(num);
export const fix_big_num = (num, place)=>{
  const l = Math.floor(Math.log10(num));
  return Math.round(num / ((10**l)/(10**place)))*((10**l)/(10**place));
}
export const format_seconds = (secs, short=true)=>{
  let neg = secs < 0;
  if (neg) secs = Math.abs(secs);
  if (secs < 60) return `${neg ? '-' : ''}${round(secs, 0)}${short ? "s" : " Seconds"}`;
  secs /= 60;
  if (secs < 60) return `${neg ? '-' : ''}${ceil(secs)}${short ? "m" : " Minutes"}`;
  secs /= 60;
  if (secs < 60) return `${neg ? '-' : ''}${ceil(secs)}${short ? "h" : " Hours"}`;
  secs /= 24;
  if (secs < 24) return `${neg ? '-' : ''}${ceil(secs)}${short ? "d" : " Days"}`;
  secs /= 365;
  return `${neg ? '-' : ''}${ceil(secs)}${short ? "y" : " Years"}`;
}
export const run_n_times = (n, callback)=>{
  Array.from(Array(n)).forEach(callback);
}
export const floor_round = (num, place)=>{
  const pow = (Math.pow(10, place));
  return Math.floor(num * pow) / pow;
}
const num_shorts = ['', 'K', 'M', 'B', 'T', 'Qa', 'Qi', 'Sx', 'Sp', 'O', 'N', 'D', 'UD', 'DD', 'TD', 'QuD', 'QiD', 'SxD', 'SpD', 'OD', 'ND', 'V', 'UV', 'DV', 'TV', 'QaV', 'QiV', 'SxV', 'SpV', 'OV', 'NV', 'T', 'UT', 'DT', 'TT', 'QaT', 'QiT', 'SxT', 'SpT', 'OT', 'NT'];
export const fnum = (num, round_to=1, i=0, past_thresh=false)=>{
    const div = num / 1000;
    const thresh = (i >= num_shorts.length);
    if (div < 1 || thresh) { 
      if (thresh) return (floor_round(num, round_to) + num_shorts[num_shorts.length-1]);
      else return (i == 0) ? (floor_round(num, round_to)) : (floor_round(num, round_to) + num_shorts[i]);
    }
    return fnum(div, round_to, i+1, thresh);
}
export const dist1 = (x1,y1, x2,y2)=>{
	const dx = x1 - x2; const dy = y1 - y2;
	return Math.sqrt( dx*dx + dy*dy );
}
export const dist2 = (pos1, pos2)=>{
	const dx = pos1.x - pos2.x, dy = pos1.y - pos2.y;
	return Math.sqrt( dx*dx + dy*dy );
}

/** c: cash | x: starting about | m: increase */
export const mult_comp = (c, x, m)=>{
  // c = x * (m^n);
  // c/x = m^n
  // log(c/x)/log(m) = n
  const pow = Math.floor(Math.log(c/x)/Math.log(m));
  const total = Math.floor(x * (m ** pow))

  return {
    pow, 
    total,
  }
}
/** c: starting cost | i: increments | x: increase */
export const add_comp = (c, i, x)=>{
  // 50 * 5 + 10 * m;
  const m = (i-1)/2 * i;
  return c * i + x * m;
}
// console.log(add_comp(50, 6, 10));

let run_n_save = null;
export const run_n = (func=()=>{}, n)=>{
  if (run_n_save == func) return;
  run_n_save = func;
  for (let i = 0; i < n; i++) func();
  run_n_save = null;
}

export const spend_cash_add = (cash, cost, incr)=>{
  let i = 0;
  if (incr == 0) {
    i = Math.floor(cash/cost);
    cash -= i*cost;
  } else {
    for ([]; cost <= cash; (cost += incr, i++)) {
      cash -= cost;
    }
  }
  return {
    cash,
    cost,
    i,
  }
}
export const spend_cash_mult = (cash, cost, incr)=>{
  let i = 0;
  if (incr == 0) {
    i = Math.floor(cash/cost);
    cash -= i*cost;
  } else {
    for ([]; cash > cost; (cost *= incr, i++)) {
      cash -= cost;
      console.log(cash);
    }
  }
  return {
    cash,
    cost,
    i,
  }
}

src/utils/entity.js
import draw from "./draw.js";
import { drop_speeds, player } from "../game.js";
import { cash } from "../stores.js";
import { dist2 } from "../utils/utils.js";

export const entity_m = (()=>{
	const list = [];
	const ids = {};
	const pool = {
		with_comp: {},
	};

	return {
		new(ent) {
			list.push(ent);
		},
		get last() { return list.at(-1); },
		update() {
			// for (let i = 0; i < list.length; i++) {
			for (let i = list.length-1; i >= 0; i--) {
				const ent = list[i];
				if (ent.destroy === true) { 
					list.splice(i, 1); 
					if (ent.id != undefined && ent.id != "none") delete ids[ent.id];
					continue; 
				}
				if (ent.enabled === false) continue;
				for (const k in ent) {
					if (!Object.hasOwnProperty.call(ent, k) || Systems[k] == undefined) continue;
					const v = ent[k];
					Systems[k](v, ent);
				}
			}
		},
		get list() { return list; },

		gid(id) {
			if (ids[id] != undefined) return ids[id]; 
			for (let i = 0; i < list.length; i++) {
				const ent = list[i];
				if (ent.id == id) {
					ids[id] = ent;
					return ent;
				}
			}
		},
		all_comps(comp) {
			let res = [];
			for (let i = 0; i < list.length; i++) {
				const ent = list[i];
				if (ent[comp] != undefined) res.push(ent[comp]);
			}
			return res;
		},
		all_with_comp(comp, and_enabled=true) {
			let res = [];
			const _list = pool.with_comp[comp] ?? list;
			for (let i = 0; i < _list.length; i++) {
				const ent = _list[i];
				if (ent[comp] != undefined) {
					if (and_enabled && ent.enabled == true) res.push(ent);
					else if (!and_enabled) res.push(ent);
				}
			}
			return res;
		},
		tag(tag) {
			const res = [];
			for (let i = 0; i < list.length; i++) {
				const ent = list[i];
				if (ent.tag == tag) res.push(ent);
			}
			return res;
		},
		match(fn=(ent)=>true) {
			const res = [];
			for (let i = 0; i < list.length; i++) {
				const ent = list[i];
				if (fn(ent) == true) res.push(ent);
			}
			return res;
		},
		pool_all_with_comp(comp) {
			pool.with_comp[comp] = this.all_with_comp(comp, false);
		},
		get pool() { return pool; },
	}
})();
export class Entity {
	constructor(options={ tag: "none", id: "none" }, enabled=true, callback=entity_m.new) {
		this.tag = options.tag ?? "none";
		this.id = options.id ?? "none";
		this.enabled = enabled;
		callback(this);
	}
	set components(list) {
		for (let i = 0; i < list.length; i++) {
			const comp = list[i]?.constructor?.name;
			if (comp == undefined) continue;
			this[comp] = list[i];
		}
	}
}

const check_overlap = (x1, y1, w1, h1, x2, y2, w2, h2)=>{
	return (
		x1 + w1 >= x2 && x1 <= x2 + w2 && y1 + h1 >= y2 && y1 <= y2 + h2
	);
}
export const closest_trig = (pos)=>{
	const ents_on = entity_m.all_with_comp("Trigger").filter(v => v.Trigger.on);
	if (ents_on.length <= 0) return null;
	if (ents_on.length == 1) return ents_on[0];
	const closest = { ent: null, dist: Infinity };
	for (let i = 0; i < ents_on.length; i++) {
		const ent = ents_on[i];
		const dist = dist2(pos, ent.Position);
		closest.dist > dist && (closest.dist = dist, closest.ent = ent);
	}
	return closest.ent;
}
const orb_lvls = [ "#ffffff", "#ff8282", "#ffb482", "#fcff82", "#a9ff82", "#82ffbb", "#82f3ff", "#82aaff", "#a782ff", "#ff82fc"];
const multiply = (axis, p1, p2)=>{
	const mults = entity_m.pool.with_comp.Upgrader.filter(ent => ent.Upgrader.axis == axis && ent.enabled == true);
	for (let i = 0; i < mults.length; i++) {
		const mult = mults[i];
		const { x, y } = mult.Position;
		if (axis == "y") {
			if (p1 < y+10 && p2 >= y+10) return 1;
		} else if (axis == "x") {
			if (p1 < x+10 && p2 >= x+10) return 1;
		}
	}
	return 0;
}

export const Components = {
	Position: class Position {
		constructor(x,y) {
			this.x = x;
			this.y = y;
		}
	},
	Graphic: class Graphic {
		constructor(draw_fn=()=>{}) {
			this.draw = draw_fn;
		}
	},
	Collider: class Collider {
		constructor(w, h, color="black") {
			this.w = w;
			this.h = h;
			this.color = color;
			this.fill = true;
			this.on_hit = ()=>{};
		}
	},
	Trigger: class Trigger {
		constructor(w=10, h=10, color={ on: "#ff6b0099", off: "#ff000099" }, hint="...") {
			this.w = w;
			this.h = h;
			this.color = {
				on: color?.on ?? "#ff6b0099",
				off: color?.off ?? "#ff000099",
			};
			this.on = false;
			this.activate = ()=>{};
			this.hint = hint;
		}
	},
	Dropper: class Dropper {
		constructor(x,y) {
			this.x = x;
			this.y = y;
			this.ticks = 0;
		}
	},
	Upgrader: class Upgrader {
		constructor(axis) {
			this.axis = axis;
		}
	},
	Orb: class Orb {
		constructor(lvl=0, val=1) {
			this.lvl = lvl;
			this.val = val;
		}
	},
	OrbManager: class OrbManager {
		constructor() {
			this.list = [];
		}
		new(orb) {
			this.list.push(orb);
		}
	}
};

export const Systems = {
	Position(pos, ent) {},
	Graphic(gr, ent) {
		if (ent.Position != undefined) draw.transform(ent.Position.x, ent.Position.y);
		gr.draw(ent);
		draw.reset();
	},
	Collider(col, ent) {
		const { w, h, on_hit } = col;
		const { x, y } = ent.Position;
		const { color, fill } = col;
		const [ pw, ph ] = [ player.size, player.size ]
		if (check_overlap(x,y,w,h, player.x,player.y,pw,ph)) {
			if (!col.hit) {
				on_hit(col);
				col.hit = true;
			}
			const to_left = player.x+pw - x;
			const to_right = x+w - player.x;
			const to_top = player.y+ph - y;
			const to_bottom = y+h - player.y;
			const closest = Math.min(to_left, to_right, to_top, to_bottom);

			if (closest == to_top) player.y = y-player.size;
			if (closest == to_bottom) player.y = y+h;
			if (closest == to_left) player.x = x-player.size;
			if (closest == to_right) player.x = x+w;
		} else if (col.hit) col.hit = false;
		ent.Graphic ?? (color != "#" && draw.rect(x, y, w, h, color, fill));
	},
	Trigger(trig, ent) {
		if (ent.Position == undefined) throw `Entities with Triggers need positions!\nId: ${ent.id}, Tag: ${ent.tag}`
		const { x, y } = ent.Position;
		const { w, h } = trig;
		const overlap = trig.on = check_overlap(x,y, w,h, player.x, player.y, player.size, player.size);
		const color = overlap && ent == closest_trig(player) ? trig.color.on : trig.color.off;
		draw.rect(x,y, w, h, color);
		draw.rect(x,y, w, h, color, false);
	},
	Dropper(drop, ent) {
		if (ent.tag == "drop_group_1") {
			draw.transform(ent.Position.x, ent.Position.y);
			draw.rect(0,-3, 30,46, "#3c3c3c");
			draw.rect(30,5, 45,30, "#3c3c3c");
			draw.reset();
		} else if (ent.tag == "drop_group_2") {
			draw.transform(ent.Position.x, ent.Position.y);
			draw.rect(5,0, 30,45, "#3c3c3c");
			draw.rect(-3,45, 46,30, "#3c3c3c");
			draw.reset();
		}

		if (0 < drop.ticks--) return;
		drop.ticks = drop_speeds[ent.tag];
		const off = (Math.round(Math.random()*15-10)); 
		const x = drop.x + (ent.tag == "drop_group_1" ? off : 0);
		const y = drop.y + (ent.tag == "drop_group_2" ? off : 0); 
		entity_m.gid("orb_m").OrbManager.new({
			x, y, off,
			lvl: 0, val: 1,
		});
	},
	Upgrader(mult, ent) {
		draw.transform(ent.Position.x, ent.Position.y);
		if (mult.axis == "y") {
			draw.rect(0,0, 5, 25, "#00aba7");
			draw.rect(5,5, 50, 15, "#00aba7");
			draw.rect(55,0, 5, 25, "#00aba7");
		} else if (mult.axis == "x") {
			draw.rect(0,0, 25, 5, "#00aba7");
			draw.rect(5,5, 15, 50, "#00aba7");
			draw.rect(0,55, 25, 5, "#00aba7");
		}
		draw.reset();
	},
	Orb(orb, ent) {
		let move_axis = "";
		if (orb.x-orb.off < 920) (orb.lvl += multiply("x", orb.x, ++orb.x), move_axis = "x");
		else if (orb.y < 510) (orb.lvl += multiply("y", orb.y, ++orb.y), move_axis = "y");
		else {
			ent.destroy = true;
			cash.update( v => v+orb.val+orb.lvl );
			return;
		}
		
		// draw.circle(orb.x, orb.y, 5, orb_lvls[orb.lvl]);
		draw.rect(orb.x-5, orb.y-5, 10, 10, orb_lvls[orb.lvl]+"66");
		
	},
	OrbManager(om, ent) {
		// for (let i = 0; i < om.list.length; i++) {
		for (let i = om.list.length-1; i >= 0; i--) {
			const orb = om.list[i];
			this.Orb(orb, orb);
			if (orb.destroy === true) om.list.splice(i, 1);
		}
	},
};

src/old_components.js
//#region | Entities
const entities = {
	list: [],
	funcs: {},
	add(data) {
		this.list.push(data);
	},
	get_all(tag="tag") {
		const query = [];
		for (let i = 0; i < this.list.length; i++) {
			const ent = this.list[i];
			if (ent.tag.includes(tag)) query.push(ent);
		}
		return query;
	},
	gid(id) {
		for (let i = 0; i < this.list.length; i++) {
			const ent = this.list[i];
			if (ent.id == id) return ent;
		}
		return null;
	},
	bind(tag="tag", func=(ent={})=>{}) {
		this.funcs[tag] = func;
	},
	update() {
		const to_dest = [];
		for (let i = 0; i < this.list.length; i++) {
			const ent = this.list[i];
			if (ent.destroy === true) {
				to_dest.push(i);
				continue;
			}
			if (ent.enabled === false) continue;
			if (ent.tag.length <= 0) return;
			for (let j = 0; j < ent.tag.split(" ").length; j++) {
				const tag = ent.tag.split(" ")[j];
				if (typeof this.funcs[tag] == "function") this.funcs[tag](ent);
			}
		}
		if (to_dest.length <= 0) return;
		to_dest.reverse()
		for (let j = 0; j < to_dest.length; j++) {
			const i = to_dest[j];
			this.list.splice(i, 1);
		}
	},
	update_order(...tags) {
		const build = [];
		for (let i = 0; i < tags.length; i++) {
			const tag = tags[i];
			build = build.concat(this.get_all(tag));
		}
		this.list = build;
	},
	apply(index, obj) {
		const ent = this.list.at(index);
		for (const k in obj) {
			if (!Object.hasOwnProperty.call(obj, k)) continue;
			const v = obj[k];
			ent[k] = v;
		}
		return ent;
	}
}
entities.bind("collider", (col)=>{
	const { x, y, w, h, on_hit } = col;
	const { color, fill } = col;
	const [ pw, ph ] = [ player.size, player.size ]
	if (check_overlap(x,y,w,h, player.x,player.y,pw,ph)) {
		if (!col.hit) {
			on_hit(col);
			col.hit = true;
		}
		const to_left = player.x+pw - x;
		const to_right = x+w - player.x;
		const to_top = player.y+ph - y;
		const to_bottom = y+h - player.y;
		const closest = Math.min(to_left, to_right, to_top, to_bottom);

		if (closest == to_top) {
			player.y = y-player.size
		}
		if (closest == to_bottom) {
			player.y = y+h;
		}
		if (closest == to_left) {
			player.x = x-player.size;
		}
		if (closest == to_right) {
			player.x = x+w;
		}
	} else if (col.hit) col.hit = false;
	draw.rect(x, y, w, h, color, fill)
});
entities.bind("dropper", (drop)=>{
	const { x, y } = drop.drop_at;
	if (orbs.ticks <= 0) {
		orbs.list.push({
			x, y,
			collect: false,
			lvl: 0,
		});
		// console.log(orbs);
	}
});
entities.bind("trigger", (trig)=>{
	const { x, y, w, h, enter, leave } = trig;
	const { color, enter_color, fill, width } = trig.style;
	if (check_overlap(x,y,w,h, player.x, player.y, player.size, player.size)) {
		if (!trig.entered) {
			enter(trig);
			trig.entered = true;
			if (typeof trig.trigger_info == "string") set_trigger_info(trig, trig.trigger_info);
		}
		// on_trigger = true;
		draw.rect(x,y,w,h, enter_color, fill, width);
	}
	else {
		if (trig.entered) {
			leave(trig);
			trig.entered = false;
			if (typeof trig.trigger_info == "string") on_trigger = false;
		}
		draw.rect(x,y,w,h, color, fill, width);
	}
});
const orb_lvl_colors = [ "#fff", "#ff8282", "#ffb482", "#fcff82", "#a9ff82", "#82ffbb", "#82f3ff", "#82aaff", "#a782ff", "#ff82fc"]
entities.bind("orbs", (orbs)=>{
	const { list } = orbs;

	for (let i = 0; i < list.length; i++) {
		const orb = list[list.length - i - 1];
		draw.circle(orb.x, orb.y, 5, orb_lvl_colors[orb.lvl]);
		if (orb.x < 920) orb.lvl += multiply("x", orb.x, ++orb.x);
		else if (orb.y < 510) orb.lvl += multiply("y", orb.y, ++orb.y);
		else {
			orb.collect = true;
			$cash += orb.lvl;
		}
	}

	// orbs.list = orbs.list.filter( orb => orb.collect == false );
	for (let i = list.length - 1; i >= 0; i--) {
		const orb = list[i];
		if (orb.collect === true) {
			orbs.list.splice(i, 1);
		}
	}

	if (orbs.ticks <= 0) {
		// list.push({
		// 	x: 920,
		// 	y: 55,
		// 	collect: false,
		// });
		orbs.ticks = 60;
		// console.log(list);
	}
	orbs.ticks--;
});
const multiply = (xy, pos1, pos2)=>{
	const mults = entities.get_all("multiplier");
	for (const mult of mults) {
		if (mult.xy != xy || mult.enabled === false) continue;
		if (pos1 <= mult[xy]+mult.h/2 && pos2 > mult[xy]+mult.h/2) {
			return 1;
		}
	}
	return 0;
}
entities.bind("multiplier", (mult)=>{

});
//#endregion
//#region | Triggers
let on_trigger = false;
/** @type HTMLElement */
let trigger_info;
const set_trigger_info = (trig, text)=>{
	on_trigger = true;
	trigger_info.style.left = trig.x+trig.w/2 + "px";
	trigger_info.style.top = trig.y+trig.h+10 + "px";
	trigger_info.innerText = text;
	// trigger_info.style.display = "block";
}
const triggers = {
	list: [],
	add(x=10,y=10, w=10,h=10, callbacks={enter:()=>{},leave:()=>{},activate:()=>{}},  style={ color: "#ff000066", enter_color: "#ff000066", fill: true, width: 1 }) {
		entities.add({ 
			tag: "trigger", x, y, w, h, 
			enter: callbacks.enter ?? (()=>{}), 
			leave: callbacks.leave ?? (()=>{}), 
			activate: callbacks.activate ?? (()=>{}),
			entered: false,
			style: {
				color: style.color ?? "#ff000066",
				enter_color: style.enter_color ?? "#ff000066",
				fill: style.fill ?? true,
				width: style.width ?? 1,
			}
		});
	},
	press_e() {
		for (const trig of entities.get_all("trigger")) {
			if (trig.entered) trig.activate(trig);
			if (trig.destroy == true || trig.enabled == false) on_trigger = false;
		}
	}
};
triggers.add(100, 100, 50, 50, { activate(t){ t.destroy = true; }, }, { color: "#ff000099" })
void entities.apply(-1, { id: "test_trigger", trigger_info: "Testing Testing 123" });

triggers.add(800, 40, 30, 30, {}, { color: "#ff000099", enter_color: "#ff6b0099" });
void entities.apply(-1, { id: "drop_trigger_1", trigger_info: "Dropper 1" });

triggers.add(800, 140, 30, 30, {activate(t) { entities.gid("dropper_2").enabled = entities.gid("drop_trigger_2").enabled = t.destroy = true }}, { color: "#ff000099", enter_color: "#ff6b0099" });
void entities.apply(-1, { id: "unlock_drop_trigger_2", trigger_info: "Unlock Dropper 2" });

triggers.add(800, 240, 30, 30, {activate(t) { entities.gid("dropper_3").enabled = entities.gid("drop_trigger_3").enabled = t.destroy = true }}, { color: "#ff000099", enter_color: "#ff6b0099" });
void entities.apply(-1, { id: "unlock_drop_trigger_3", trigger_info: "Unlock Dropper 3" });

triggers.add(800, 140, 30, 30, {}, { color: "#ff000099", enter_color: "#ff6b0099" });
void entities.apply(-1, { id: "drop_trigger_2", trigger_info: "Dropper 2", enabled: false });

triggers.add(800, 240, 30, 30, {}, { color: "#ff000099", enter_color: "#ff6b0099" });
void entities.apply(-1, { id: "drop_trigger_3", trigger_info: "Dropper 3", enabled: false });

triggers.add(800, 340, 30, 30, {activate(t) { entities.gid("mult_1").enabled = t.destroy = entities.gid("unlock_mult_trigger_2").enabled = true }}, { color: "#ff000099", enter_color: "#ff6b0099" });
void entities.apply(-1, { id: "unlock_mult_trigger_1", trigger_info: "Unlock Multiplier 1" });

triggers.add(800, 340, 30, 30, {activate(t) { entities.gid("mult_2").enabled = t.destroy = entities.gid("unlock_mult_trigger_3").enabled = true; }}, { color: "#ff000099", enter_color: "#ff6b0099" });
void entities.apply(-1, { id: "unlock_mult_trigger_2", trigger_info: "Unlock Multiplier 2", enabled: false });

triggers.add(800, 340, 30, 30, {activate(t) { entities.gid("mult_3").enabled = t.destroy = true; }}, { color: "#ff000099", enter_color: "#ff6b0099" });
void entities.apply(-1, { id: "unlock_mult_trigger_3", trigger_info: "Unlock Multiplier 3", enabled: false });
//#endregion
//#region | Colliders
const colliders = {
	list: [],
	add(x=10,y=10,w=10,h=10, on_hit=(col)=>{}, extra={ color: "#0000ff55", fill: true, width: 1 }) {
		entities.add({ 
			tag: "collider", 
			x, y, w, h, on_hit, hit: false,
			color: extra.color ?? "#00000000",
			fill: extra.fill ?? true,
			width: extra.width ?? 1,
		});
	}
}
colliders.add(860, 510, 120, 70, ()=>{}, { color: "#222" } );
void entities.apply(-1, { id: "collector_block" });

colliders.add(895, 510, 50, 10, ()=>{}, { color: "#444" } );
void entities.apply(-1, { id: "collector_block_door" });

colliders.add(895, 20, 50, 490, ()=>{}, { color: "#555" } );
void entities.apply(-1, { id: "conveyor_1" });

entities.add({
	tag: "orbs",
	list: [],
	ticks: 0,
});
const orbs = entities.list.at(-1);

colliders.add(835, 40, 100, 30, ()=>{}, { color: "#444" } );
void entities.apply(-1, { id: "dropper_1", tag: "collider dropper", drop_at: { x: 920, y: 55 } });

colliders.add(835, 140, 100, 30, ()=>{}, { color: "#444" } );
void entities.apply(-1, { id: "dropper_2", tag: "collider dropper", enabled: false, drop_at: { x: 920, y: 155 }  });

colliders.add(835, 240, 100, 30, ()=>{}, { color: "#444" } );
void entities.apply(-1, { id: "dropper_3", tag: "collider dropper", enabled: false, drop_at: { x: 920, y: 255 }  });

colliders.add(890, 350, 60, 15, ()=>{}, { color: "#444" } ); void entities.apply(-1, { id: "mult_1", tag: "collider multiplier", xy: "y", enabled: false });
colliders.add(890, 400, 60, 15, ()=>{}, { color: "#444" } ); void entities.apply(-1, { id: "mult_2", tag: "collider multiplier", xy: "y", enabled: false });
colliders.add(890, 450, 60, 15, ()=>{}, { color: "#444" } ); void entities.apply(-1, { id: "mult_3", tag: "collider multiplier", xy: "y", enabled: false });
//#endregion

